src/server/JasmineGraphInstanceService.cpp

```diff
@@ -614,62 +614,54 @@ void *instanceservicesession(void *dummyPt) {
+            bzero(data, INSTANCE_DATA_LENGTH);
+            read(connFd, data, INSTANCE_DATA_LENGTH);
+            string graphID = (data);
+            graphID = utils.trim_copy(graphID, " \f\n\r\t\v");
+            instance_logger.log("Received Graph ID: " + graphID, "info");
 
-            for (its = degreeDistributionCentral.begin(); its != degreeDistributionCentral.end(); ++its) {
-                bool centralNodeFound = false;
-                for (itcentral = degreeDistribution.begin(); itcentral != degreeDistribution.end(); ++itcentral) {
-                    if ((its->first) == (itcentral->first)) {
-                        degreeDistribution[its->first] = (its->second) + (itcentral->second);
-                        centralNodeFound = true;
-                        break;
-                    }
-                }
+            write(connFd, JasmineGraphInstanceProtocol::OK.c_str(), JasmineGraphInstanceProtocol::OK.size());
+            instance_logger.log("Sent : " + JasmineGraphInstanceProtocol::OK, "info");
 
-                if (!centralNodeFound) {
-                    degreeDistribution.insert(std::make_pair(its->first, its->second));
-                }
-            }
+            bzero(data, INSTANCE_DATA_LENGTH);
+            read(connFd, data, INSTANCE_DATA_LENGTH);
+            string partitionID = (data);
+            partitionID = utils.trim_copy(partitionID, " \f\n\r\t\v");
+            instance_logger.log("Received Partition ID: " + partitionID, "info");
 
-            string inDegreeDistString;
-            int count = 0;
-            for (its = degreeDistribution.begin(); its != degreeDistribution.end(); ++its) {
-                count++;
-                inDegreeDistString.append(std::to_string(its->first) + ":" + std::to_string(its->second) + ",");
 
-                if (count == JasmineGraphInstanceService::MESSAGE_SIZE) {
-                    write(connFd, inDegreeDistString.c_str(), inDegreeDistString.size());
-                    inDegreeDistString = "";
-                    count = 0;
-                }
-            }
+            map<long, long> degreeDistribution = calculateLocalInDegreeDist(graphID, partitionID,
+                                                                            graphDBMapLocalStores,
+                                                                            graphDBMapCentralStores);
+            instance_logger.log("In Degree Dist size: " + to_string(degreeDistribution.size()), "info");
 
-            write(connFd, JasmineGraphInstanceService::END_OF_MESSAGE.c_str(),
-                  JasmineGraphInstanceService::END_OF_MESSAGE.size());
-            instance_logger.log("Sent : " + JasmineGraphInstanceService::END_OF_MESSAGE, "info");
-        } else if (line.compare(JasmineGraphInstanceProtocol::DP_CENTRALSTORE) == 0) {
+            //todo kasundharmadasa : The in degree distribution of a worker is kept in the worker itself
```

Comment:
Do we keep the in-degree distribution of each partition in memory? or do we store them in secondary storage?

Reply:
Ok